hpy/devel/src/runtime/hlist.c

```diff
@@ -0,0 +1,135 @@
+/**
+ * A manager for a list of HPy handles, allowing handles to be tracked
+ * and closed as a group.
+ *
+ * Note::
+ *    HList always keeps space for one extra handle free so that
+ *    HList_Track can always store the handle being passed to it,
+ *    even if it fails to automatically create space for future
+ *    handles. This allows HList_Free to close all handles passed to
+ *    HList_Track.
+ *
+ *    As a consequence, the minimum size for an HList is one handle.
+ *
+ * Example usage (inside an HPyDef_METH function)::
+ *
+ * long i;
+ * HPy key, value;
+ * HList hl;
+ *
+ * hl = HList_New(ctx);  // track the key-value pairs
+ * if (hl == NULL)
+ *     return HPy_NULL;
+ *
+ * HPy dict = HPyDict_New(ctx);
+ * if (HPy_IsNull(dict))
+ *     goto error;
+ *
+ * for (i=0; i<5; i++) {
+ *     key = HPyLong_FromLong(ctx, i);
+ *     if (HPy_IsNull(key))
+ *         goto error;
+ *     if (HList_Track(ctx, hl, key) < 0)
+ *         goto error;
+ *     value = HPyLong_FromLong(ctx, i * i);
+ *     if (HPy_IsNull(value)) {
+ *         goto error;
+ *     }
+ *     if (HList_Track(ctx, hl, value) < 0)
+ *         goto error;
+ *     result = HPy_SetItem(ctx, dict, key, value);
+ *     if (result < 0)
+ *         goto error;
+ * }
+ *
+ * success:
+ *    HList_Free(ctx, hl);
+ *    return dict;
+ *
+ * error:
+ *    HList_Free(ctx, hl);
+ *    HPy_Close(ctx, dict);
+ *    HPyErr_SetString(ctx, ctx->h_ValueError, "Failed!");
+ *    return HPy_NULL;
+ */
+
+#include <Python.h>
```

Comment:
If you put it in `devel/src/runtime` you are NOT allowed to include `Python.h`: these files are automatically included in all `hpy_ext_modules`, including the universal ones.
So, two solutions:
1. keep it in `hpy/devel/src/runtime` but implement everything in terms of other HPy primitives (as `argparse.c`)
2. move it to `hpy/universal`, and let each implementation to write its own version. In this case all functions like `HList_New` & co. will probably need to go inside `public_api.h`